### PR TITLE
feat: expose current property name to transform function

### LINKED ISSUE #329: feat: pass property name to a handler of @Transform
Let's pass property name to a handler of @Transform to give developers more flexibility.

Use case:
```ts
import { ObjectId } from 'mongodb';

export class CreateRecordDto {
  @Type(() => ObjectId)
  @Transform(getObjectIdTransformer('clientId'))
  readonly clientId: ObjectId;
}

function getObjectIdTransformer(propertyName: string) {
  return (value: string) => {
    if (!ObjectId.isValid(value)) {
      throw new HttpException(`Property ${propertyName} is an invalid MongoDB ObjectId`, HttpStatus.UNPROCESSABLE_ENTITY);
    }
    return new ObjectId(value);
  };
}
```

To construct exception message I need to use function-wrapper and specify a property name. In application I have a few places where I need to use custom transformation for different properties such as "clientId", "masterId", "id", etc.

Will be nice to know a handled property name inside transform function (handler).

Desired result is something like this:
```ts
export class CreateRecordDto {
  @Type(() => ObjectId)
  @Transform(objectIdTransformer)
  readonly clientId: ObjectId;
}

function objectIdTransformer(value: string, propertyName: string) {
  if (!ObjectId.isValid(value)) {
    throw new HttpException(`Property ${propertyName} is an invalid MongoDB ObjectId`, HttpStatus.UNPROCESSABLE_ENTITY);
  }
  return new ObjectId(value);
}
```

From my point of view new argument `key` should be the second to have the same order as for other Javascript callback handlers such as for `[].map(value, index, arr)`, `[].forEach(value, index, arr)`, etc.

### PR DESCRIPTION
**Pull request by pakhuta**
## Description

This PR is a continuation of the https://github.com/typestack/class-transformer/pull/330 (because fork was deleted).

- updated the signature of the transform function: `transformFn({ value, key, object, type })`
- added additional checks to an existed test case

## Checklist

- [x] the pull request title describes what this PR does (not a vague title like `Update index.md`)
- [x] the pull request targets the *default* branch of the repository (`develop`)
- [x] the code follows the established code style of the repository
  - `npm run prettier:check` passes
  - `npm run lint:check` passes
- [x] tests are added for the changes I made (if any source code was modified)
- [ ] documentation added or updated (no, due to the comment of "parent" [PR](https://github.com/typestack/class-transformer/pull/330#issuecomment-665826602) )

> remove all changes not related to the code (eg: changelog change, version number)

- [x] I have run the project locally and verified that there are no errors

### Fixes

fixes #329

### CODE REVIEW

**Comment by NoNameProvided** [comment]
Thanks! I will take a look soon!

File: src/interfaces/metadata/transform-metadata.interface.ts [hunk 1]
```diff
@@ -15,10 +15,17 @@ export interface TransformMetadata {
...
-  transformFn: (value: any, obj: any, transformationType: TransformationType) => any;
+  transformFn: (params: TransformFnParams) => any;
 
   /**
    * Options passed to the @Transform operator for this property.
    */
   options: TransformOptions;
 }
+
+export interface TransformFnParams {
```
Comment (NoNameProvided) [inline src/interfaces/metadata/transform-metadata.interface.ts:26]: Please move the interface to its own file under `src/interfaces/metadata/transform-fn-params.interface.ts`.

**Comment by NoNameProvided** [comment]
Thank you!

**Comment by pakhuta** [comment]
😀 Thank you for the package!

**Pull request merged**